libp2p/NodeTable.h

```diff
@@ -0,0 +1,363 @@
+/*
+ This file is part of cpp-ethereum.
+ 
+ cpp-ethereum is free software: you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+ 
+ cpp-ethereum is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+ 
+ You should have received a copy of the GNU General Public License
+ along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
+ */
+/** @file NodeTable.h
+ * @author Alex Leverington <nessence@gmail.com>
+ * @date 2014
+ */
+
+#pragma once
+
+#include <algorithm>
+#include <boost/integer/static_log2.hpp>
+#include <libdevcrypto/Common.h>
+#include <libp2p/UDP.h>
+
+namespace dev
+{
+namespace p2p
+{
+
+/**
+ * NodeTable using S/Kademlia system for node discovery and preference.
+ * untouched buckets are refreshed if they have not been touched within an hour
+ *
+ * Thread-safety is ensured by modifying NodeEntry details via 
+ * shared_ptr replacement instead of mutating values.
+ *
+ * [Integration]
+ * @todo deadline-timer which maintains tcp/peer connections
+ * @todo restore nodes: affects refreshbuckets
+ * @todo TCP endpoints
+ * @todo makeRequired: don't try to evict node if node isRequired.
+ * @todo makeRequired: exclude bucket from refresh if we have node as peer.
+ *
+ * [Optimization]
+ * @todo Pong to include ip:port where ping was received
+ * @todo expiration and sha3(id) 'to' for messages which are replies (prevents replay)
+ * @todo std::shared_ptr<PingNode> m_cachedPingPacket;
+ * @todo std::shared_ptr<FindNeighbors> m_cachedFindSelfPacket;
+ *
+ * [Networking]
+ * @todo TCP endpoints
+ * @todo eth/upnp/natpmp/stun/ice/etc for public-discovery
+ * @todo firewall
+ *
+ * [Protocol]
+ * @todo post-eviction pong
+ * @todo optimize knowledge at opposite edges; eg, s_bitsPerStep lookups. (Can be done via pointers to NodeBucket)
+ * @todo ^ s_bitsPerStep = 5; // Denoted by b in [Kademlia]. Bits by which address space is divided.
+ * @todo optimize (use tree for state and/or custom compare for cache)
+ * @todo reputation (aka universal siblings lists)
+ */
+class NodeTable: UDPSocketEvents, public std::enable_shared_from_this<NodeTable>
+{
+	friend struct Neighbors;
+	using NodeSocket = UDPSocket<NodeTable, 1280>;
+	using TimePoint = std::chrono::steady_clock::time_point;
+	using EvictionTimeout = std::pair<std::pair<Address,TimePoint>,Address>;
+
+	struct NodeDefaultEndpoint
+	{
+		NodeDefaultEndpoint(bi::udp::endpoint _udp): udp(_udp) {}
```

Comment:
This struct will hold additional information and the node may have other endpoints thanks to ipv6.